pkg/services/object_manager/placement/traverser.go

```diff
@@ -0,0 +1,182 @@
+package placement
+
+import (
+	"sync"
+
+	"github.com/nspcc-dev/neofs-api-go/pkg/container"
+	"github.com/nspcc-dev/neofs-api-go/pkg/netmap"
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+	"github.com/nspcc-dev/neofs-node/pkg/network"
+	"github.com/pkg/errors"
+)
+
+// Builder is an interface of the
+// object placement vector builder.
+type Builder interface {
+	// BuildPlacement returns the list of placement vectors
+	// for object according to the placement policy.
+	//
+	// Must return all container nodes if object identifier
+	// is nil.
+	BuildPlacement(*object.Address, *netmap.PlacementPolicy) ([]netmap.Nodes, error)
+}
+
+// Option represents placement traverser option.
+type Option func(*cfg)
+
+// Traverser represents utility for controlling
+// traversal of object placement vectors.
+type Traverser struct {
+	mtx *sync.RWMutex
+
+	rem int
+
+	nextI, nextJ int
+
+	vectors []netmap.Nodes
+}
+
+type cfg struct {
+	rem int
+
+	addr *object.Address
+
+	policy *netmap.PlacementPolicy
+
+	builder Builder
+}
+
+var errNilBuilder = errors.New("placement builder is nil")
+
+func defaultCfg() *cfg {
+	return &cfg{
+		rem:  1,
+		addr: object.NewAddress(),
+	}
+}
+
+// NewTraverser creates, initializes with options and returns Traverser instance.
+func NewTraverser(opts ...Option) (*Traverser, error) {
+	cfg := defaultCfg()
+
+	for i := range opts {
+		if opts[i] != nil {
+			opts[i](cfg)
+		}
+	}
+
+	if cfg.builder == nil {
+		return nil, errors.Wrap(errNilBuilder, "incomplete traverser options")
+	}
+
+	ns, err := cfg.builder.BuildPlacement(cfg.addr, cfg.policy)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not build placement")
+	}
+
+	return &Traverser{
+		mtx:     new(sync.RWMutex),
+		rem:     cfg.rem,
+		vectors: ns,
+	}, nil
+}
+
+// Next returns next unprocessed address of the object placement.
+//
+// Returns nil if no nodes left or traversal operation succeeded.
+func (t *Traverser) Next() *network.Address {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	if t.rem == 0 || t.nextI == len(t.vectors) {
+		return nil
+	}
+
+	addr, err := network.AddressFromString(t.vectors[t.nextI][t.nextJ].NetworkAddress())
+	if err != nil {
+		// TODO: log error
+	}
+
+	if t.nextJ++; t.nextJ == len(t.vectors[t.nextI]) {
```

Comment:
Is this going to work correctly if two vectors have different length?

Reply:
It should work with any two-dimensional slice. Do you see any problems here?

Reply:
Okay, I see now how it works.

Another question: what if there is a placement rule like 
```
PUT 2 in HDD
PUT 1 in SSD
...
SELECT 6 Nodes .. As HDD
SELECT 4 Nodes .. As SSD
```
Does traverser work with single replica statement? In this case traverser will be created twice, one for HDD with `t.rem == 2` and one for SSD with `t.rem == 1`.